src/target/lpc55xx.c

```diff
@@ -37,12 +37,478 @@
+	case LPC5528_CHIPID:
+		return "LPC5528";
+	case LPC55S28_CHIPID:
+		return "LPC55S28";
+
+	default:
+		return "unknown";
+	}
+}
+
+static int lpc55xx_get_rom_api_version(target* t, target_addr_t bootloader_tree_address)
+{
+	return (((target_mem_read32(t, bootloader_tree_address + 0x4) >> 16) & 0xff) == 3) ? 1 : 0;
+}
+
+static target_addr_t lpc55xx_get_flash_table_address(target *t, target_addr_t bootloader_tree_address)
+{
+	return target_mem_read32(t, bootloader_tree_address + 0x10);
+}
+
+static target_addr_t lpc55xx_get_flash_init_address(target* t)
+{
+	target_addr_t bootloader_tree_address = lpc55xx_get_bootloader_tree_address(t);
+
+	target_addr_t flash_table_address = lpc55xx_get_flash_table_address(t, bootloader_tree_address);
+	return target_mem_read32(t, flash_table_address + 1 * sizeof(uint32_t));
+}
+
+static target_addr_t lpc55xx_get_flash_erase_address(target* t)
+{
+	target_addr_t bootloader_tree_address = lpc55xx_get_bootloader_tree_address(t);
+
+	if (lpc55xx_get_rom_api_version(t, bootloader_tree_address) == 0) {
+		return 0x1300413bU; // UNTESTED: found in SDK, not in the UM
+	}
+
+	target_addr_t flash_table_address = lpc55xx_get_flash_table_address(t, bootloader_tree_address);
+	return target_mem_read32(t, flash_table_address + 2 * sizeof(uint32_t));
+}
+
+static target_addr_t lpc55xx_get_flash_program_address(target* t)
+{
+	target_addr_t bootloader_tree_address = lpc55xx_get_bootloader_tree_address(t);
+
+	if (lpc55xx_get_rom_api_version(t, bootloader_tree_address) == 0) {
+		return 0x1300419dU; // UNTESTED: found in SDK, not in the UM
+	}
+
+	target_addr_t flash_table_address = lpc55xx_get_flash_table_address(t, bootloader_tree_address);
+	return target_mem_read32(t, flash_table_address + 3 * sizeof(uint32_t));
+}
+
+static target_addr_t lpc55xx_get_ffr_init_address(target* t)
+{
+	target_addr_t bootloader_tree_address = lpc55xx_get_bootloader_tree_address(t);
+	target_addr_t flash_table_address = lpc55xx_get_flash_table_address(t, bootloader_tree_address);
+
+	if (lpc55xx_get_rom_api_version(t, bootloader_tree_address) == 0) {
+		return target_mem_read32(t, flash_table_address + 7 * sizeof(uint32_t));
+	} else {
+		return target_mem_read32(t, flash_table_address + 10 * sizeof(uint32_t));
+	}
+}
+
+static target_addr_t lpc55xx_get_ffr_get_uuid_address(target* t)
+{
+	target_addr_t bootloader_tree_address = lpc55xx_get_bootloader_tree_address(t);
+	target_addr_t flash_table_address = lpc55xx_get_flash_table_address(t, bootloader_tree_address);
+
+		if (lpc55xx_get_rom_api_version(t, bootloader_tree_address) == 0) {
+		return target_mem_read32(t, flash_table_address + 10 * sizeof(uint32_t));
+	} else {
+		return target_mem_read32(t, flash_table_address + 13 * sizeof(uint32_t));
+	}
+}
+
+static enum iap_status iap_call_raw(target *t, enum iap_cmd cmd, uint32_t r1, uint32_t r2, uint32_t r3)
+{
+	/* Prepare the registers for the IAP call. R0 is always flash_config */
+
+	uint32_t regs[t->regs_size / sizeof(uint32_t)];
```

Comment:
If t->regs_size is a constant size for this processor series, please use that constant here instead of this. This constructs a VLA and we're trying to remove all variable length arrays from the code base.

Reply:
What's the right way to do this? It looks like they are all M33 with FPU, so I think it would be `sizeof(regnum_cortex_m) + sizeof(regnum_cortex_mf)` bytes, but these are defined in `cortexm.c`. Should I just hardcode it to 20 + 33 = 53? I couldn't find any other example of register saving that doesn't use VLAs.

Reply:
I am unable to complete this item because the macros discussed in the discord (to retrieve the number of Cortex-M33 registers) are not available yet.

Reply:
We were suggesting this PR would include a patch to add those macros - but if you wish us to do that first, then no worries, we can do that.

Reply:
Oh sorry I didn't understand that you intended this patch to do it. I figured it was unrelated work that could be separate to this work. My bad for this miscommunication; I'll let you do it and will update the PR afterwards

Reply:
ok, works for us. We'll add the macros either to the base branch or to main once v1.9 is out (probably the base branch for this for now so you can finish this PR)

Reply:
We finally got back around to fixing this and are introducing a new Cortex header that has stuff common across the CPU family, including these much awaited register count macros